### PR TITLE
An example auditor/verifier for the Go SumDB

### PR DESCRIPTION
**Pull request by mhutchinson**

### CODE REVIEW

File: sumdbaudit/audit/database.go [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+
+// SetLeafMetadata sets the metadata for a contiguous batch of leaves.
+func (d *Database) SetLeafMetadata(ctx context.Context, start int64, metadata []Metadata) error {
+	tx, err := d.db.BeginTx(ctx, nil)
+	if err != nil {
+		return fmt.Errorf("BeginTx: %v", err)
+	}
+	for mi, m := range metadata {
+		midx := int64(mi) + start
+		tx.Exec("INSERT INTO leafMetadata (id, module, version, fileshash, modhash) VALUES (?, ?, ?, ?, ?)", midx, m.module, m.version, m.repoHash, m.modHash)
```
Comment (AlCutter) [inline sumdbaudit/audit/database.go:90]: just wondering from the code here if it's worth changing the `fileshash` column to be named `repohash` (or viceversa)?

File: sumdbaudit/audit/database.go [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+	var res []byte
+	err := d.db.QueryRow("SELECT hashes FROM tiles WHERE height=? AND level=? AND offset=?", height, level, offset).Scan(&res)
+	if err != nil {
+		return nil, err
+	}
+	return SplitTile(res, height), nil
+}
+
+// SetTile sets the leaf hash data for the given tile.
+// The leaf hashes should be 2^height * HashLenBytes long.
```
Comment (AlCutter) [inline sumdbaudit/audit/database.go:106]: Worth enforcing this invariant?

File: sumdbaudit/audit/service.go [hunk 3]
```diff
@@ -0,0 +1,361 @@
...
+	}
+}
+
+// CloneLeafTiles copies the leaf data from the SumDB into the local database.
+// It only copies whole tiles, which means that some stragglers may not be
+// copied locally.
+func (s *Service) CloneLeafTiles(ctx context.Context, checkpoint *tlog.Tree) error {
+	head, err := s.localDB.GetHead()
+	if err != nil {
+		// failed to find head of database, assuming empty and starting from scratch
```
Comment (AlCutter) [inline sumdbaudit/audit/service.go:50]: nit: that might be a good thing to `log.Info()`?

**Review by AlCutter — COMMENTED** [review]
Gotta stop here for now, I thought I'd push the nitty comments/bits I've reviewed so far in case they're helpful.

File: sumdbaudit/audit/database.go [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+	}
+	for mi, m := range metadata {
+		midx := int64(mi) + start
+		tx.Exec("INSERT INTO leafMetadata (id, module, version, fileshash, modhash) VALUES (?, ?, ?, ?, ?)", midx, m.module, m.version, m.repoHash, m.modHash)
+	}
+	return tx.Commit()
+}
+
+// GetTile gets the leaf hashes for the given tile, or returns an error.
+func (d *Database) GetTile(height, level, offset int) ([][]byte, error) {
```
Comment (AlCutter) [inline sumdbaudit/audit/database.go:96]: -> `Tile`

File: sumdbaudit/audit/service.go [hunk 5]
```diff
@@ -0,0 +1,361 @@
...
+	for level := s.getLevelsForLeafCount(checkpoint.N); level >= 0; level-- {
+		// how many real leaves a tile at this level covers.
+		tileLeafCount := uint64(1) << ((level + 1) * s.height)
+		levelTileCount := int(uint64(checkpoint.N) / tileLeafCount)
+		firstTileOffset := int(logRange.End() / tileLeafCount)
+
+		for offset := firstTileOffset; offset < levelTileCount; offset++ {
+			tHashes, err := s.localDB.GetTile(s.height, level, offset)
+			if err != nil {
+				return fmt.Errorf("failed to get tile L=%d, O=%d: %v (did you run hashfill?)", level, offset, err)
```
Comment (AlCutter) [inline sumdbaudit/audit/service.go:152]: What's `hashfill`?

File: sumdbaudit/audit/service.go [hunk 6]
```diff
@@ -0,0 +1,361 @@
...
+	var rootHash tlog.Hash
+	copy(rootHash[:], root)
+	if rootHash != checkpoint.Hash {
+		return fmt.Errorf("log root mismatch at tree size %d; calculated %x, SumDB says %x", checkpoint.N, root, checkpoint.Hash[:])
+	}
+	return nil
+}
+
+// VerifyTiles checks that every tile calculated locally matches the result returned
+// by SumDB. This shouldn't be possible if CheckRootHash is working, but this may be
```
Comment (AlCutter) [inline sumdbaudit/audit/service.go:197]: /possible/necessary/?

File: sumdbaudit/audit/service.go [hunk 7]
```diff
@@ -0,0 +1,361 @@
...
+			metadata[i] = Metadata{module, version, repoHash, modHash}
+		}
+		if err := s.localDB.SetLeafMetadata(ctx, leafOffset, metadata); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (s *Service) hashLeafLevel(tileCount int, roots chan *compact.Range) error {
```
Comment (AlCutter) [inline sumdbaudit/audit/service.go:265]: nit: probably should define the channel as `roots chan<- *compactRange` if you're only writing to it.

(same elsewhere)

File: sumdbaudit/audit/service.go [hunk 8]
```diff
@@ -0,0 +1,361 @@
...
+	topLevel := -1
+	coveredIdx := leaves >> s.height
+	for coveredIdx > 0 {
+		coveredIdx = coveredIdx >> s.height
+		topLevel++
+	}
+	return topLevel
+}
+
+type tileLeaves struct {
```
Comment (AlCutter) [inline sumdbaudit/audit/service.go:358]: Might be worth a small comment on the struct?

File: sumdbaudit/audit/sumdb.go [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+	"strings"
+
+	"golang.org/x/mod/sumdb/note"
+	"golang.org/x/mod/sumdb/tlog"
+)
+
+// HashLenBytes is the number of bytes in the SumDB hashes.
+const HashLenBytes = 32
+
+const pathBase = 1000
```
Comment (AlCutter) [inline sumdbaudit/audit/sumdb.go:18]: maybe a comment on what this is?

File: sumdbaudit/audit/sumdb.go [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+	return dataToLeaves(data), nil
+}
+
+// PartialLeavesAtOffset gets the final tile of incomplete leaves.
+func (c *SumDBClient) PartialLeavesAtOffset(offset, count int) ([][]byte, error) {
+	nStr := fmt.Sprintf("%03d", offset%pathBase)
+	for offset >= pathBase {
+		offset /= pathBase
+		nStr = fmt.Sprintf("x%03d/%s", offset%pathBase, nStr)
+	}
```
Comment (AlCutter) [inline sumdbaudit/audit/sumdb.go:93]: is it worth pulling these out and share between this and the func above?

File: sumdbaudit/audit/sumdb.go [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+	return result
+}
+
+// TileHashes gets the hashes at the given level and offset.
+func (c *SumDBClient) TileHashes(level, offset int) ([]tlog.Hash, error) {
+	nStr := fmt.Sprintf("%03d", offset%pathBase)
+	for offset >= pathBase {
+		offset /= pathBase
+		nStr = fmt.Sprintf("x%03d/%s", offset%pathBase, nStr)
+	}
```
Comment (AlCutter) [inline sumdbaudit/audit/sumdb.go:122]: here too?

**Review by AlCutter — COMMENTED** [review]
few more bits'n'bobs

File: sumdbaudit/audit/sumdb.go (hunk 10 above)
Comment (mhutchinson) [reply sumdbaudit/audit/sumdb.go:102]: Done.

File: sumdbaudit/audit/sumdb.go (hunk 11 above)
Comment (mhutchinson) [reply sumdbaudit/audit/sumdb.go]: Wow, I'm lazy. Ta.

**Review by mhutchinson — COMMENTED** [review]
Thanks Al. I've addressed your comments. And it still works.

File: sumdbaudit/audit/service.go (hunk 5 above)
Comment (AlCutter) [reply sumdbaudit/audit/service.go]: Awesomes, call me Indy :)

**Review by AlCutter — APPROVED** [review]
Nice one, ta!

BTW - looks like it's missing a licence header: `./sumdbaudit/cli/clone/clone.go:10:license header not found`

**Pull request merged**